// checkly/dist/constructs/alert-channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlertChannel = exports.AlertChannelRef = void 0;
const construct_1 = require("./construct");
const project_1 = require("./project");
/**
 * Creates a reference to an existing Alert Channel.
 *
 * References link existing resources to a project without managing them.
 */
class AlertChannelRef extends construct_1.Construct {
    constructor(logicalId, physicalId) {
        super(AlertChannel.__checklyType, logicalId, physicalId, false);
        project_1.Session.registerConstruct(this);
    }
    synthesize() {
        return null;
    }
}
exports.AlertChannelRef = AlertChannelRef;
/**
 * Creates an Alert Channels
 *
 * @remarks
 *
 * This class make use of the Alert Channels endpoints.
 */
class AlertChannel extends construct_1.Construct {
    sendRecovery;
    sendFailure;
    sendDegraded;
    sslExpiry;
    sslExpiryThreshold;
    static __checklyType = 'alert-channel';
    /**
     * Constructs the Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props alert channel configuration properties
     */
    constructor(logicalId, props) {
        super(AlertChannel.__checklyType, logicalId);
        this.sendRecovery = props.sendRecovery;
        this.sendFailure = props.sendFailure;
        this.sendDegraded = props.sendDegraded;
        this.sslExpiry = props.sslExpiry;
        this.sslExpiryThreshold = props.sslExpiryThreshold;
    }
    static fromId(id) {
        return new AlertChannelRef(`alert-channel-${id}`, id);
    }
    allowInChecklyConfig() {
        return true;
    }
    synthesize() {
        return {
            sendRecovery: this.sendRecovery,
            sendFailure: this.sendFailure,
            sendDegraded: this.sendDegraded,
            sslExpiry: this.sslExpiry,
            sslExpiryThreshold: this.sslExpiryThreshold,
        };
    }
}
exports.AlertChannel = AlertChannel;
