// checkly/dist/constructs/incidentio-alert-channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IncidentioAlertChannel = void 0;
const webhook_alert_channel_1 = require("./webhook-alert-channel");
/**
 * Creates an Incident.io Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
class IncidentioAlertChannel extends webhook_alert_channel_1.WebhookAlertChannel {
    static DEFAULT_PAYLOAD = `{
  "title": "{{ALERT_TITLE}}",
  "description": "{{ALERT_TITLE}} at {{STARTED_AT}} in {{RUN_LOCATION}} {{RESPONSE_TIME}}ms",
  "deduplication_key": "{{CHECK_ID}}",
  "metadata": {
    "alertType": "{{ALERT_TYPE}}",
    "check_result_id": "{{CHECK_RESULT_ID}}",
    "resultLink": "{{RESULT_LINK}}"
  },
  {{#contains ALERT_TYPE "RECOVERY"}}
  "status": "resolved"
  {{else}}
  "status": "firing"
  {{/contains}}
}
`;
    /**
     * Constructs the Incident.io Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props Incident.io alert channel configuration properties
     *
     * {@link https://checklyhq.com/docs/cli/constructs-reference/#incidentioalertchannel Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.webhookType = 'WEBHOOK_INCIDENTIO';
        this.method = 'POST';
        this.headers = [
            {
                key: 'authorization',
                value: `Bearer ${props.apiKey}`,
                locked: false,
            },
        ];
        this.template = props.payload ?? IncidentioAlertChannel.DEFAULT_PAYLOAD;
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'WEBHOOK',
            config: {
                name: this.name,
                webhookType: this.webhookType,
                url: this.url,
                template: this.template,
                method: this.method,
                headers: this.headers,
                queryParameters: this.queryParameters,
                webhookSecret: this.webhookSecret,
            },
        };
    }
}
exports.IncidentioAlertChannel = IncidentioAlertChannel;
