// checkly/dist/constructs/incidentio-alert-channel-codegen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IncidentioAlertChannelCodegen = void 0;
const codegen_1 = require("./internal/codegen");
const sourcegen_1 = require("../sourcegen");
const alert_channel_codegen_1 = require("./alert-channel-codegen");
const incidentio_alert_channel_1 = require("./incidentio-alert-channel");
function apiKeyFromHeaders(headers) {
    for (const header of headers) {
        if (header.key.toLocaleLowerCase() !== 'authorization') {
            continue;
        }
        if (!header.value.startsWith('Bearer ')) {
            continue;
        }
        return header.value.replace('Bearer ', '');
    }
}
const construct = 'IncidentioAlertChannel';
class IncidentioAlertChannelCodegen extends codegen_1.Codegen {
    validateSafety(resource) {
        const { config } = resource;
        if (config.method !== 'POST') {
            throw new codegen_1.ImportSafetyViolation(`Unsupported value for property 'method' (expected 'POST')`);
        }
        if (config.headers === undefined) {
            throw new codegen_1.ImportSafetyViolation(`Unsupported value for property 'headers' (expected a single 'authorization' header)`);
        }
        if (config.headers.length !== 1) {
            throw new codegen_1.ImportSafetyViolation(`Unsupported value for property 'headers' (expected a single 'authorization' header)`);
        }
        if (config.headers[0].key.toLowerCase() !== 'authorization') {
            throw new codegen_1.ImportSafetyViolation(`Unsupported value for property 'headers' (expected a single 'authorization' header)`);
        }
        if (config.queryParameters !== undefined && config.queryParameters.length !== 0) {
            throw new codegen_1.ImportSafetyViolation(`Unsupported value for property 'queryParameters' (expected no value or an empty array)`);
        }
        if (config.webhookSecret) {
            throw new codegen_1.ImportSafetyViolation(`Unsupported value for property 'webhookSecret' (expected no value)`);
        }
    }
    describe(resource) {
        this.validateSafety(resource);
        return `Incident.io Alert Channel: ${resource.config.name}`;
    }
    prepare(logicalId, resource, context) {
        this.validateSafety(resource);
        const { name } = resource.config;
        const filename = context.filePath('resources/alert-channels/incident-io', name, {
            unique: true,
        });
        context.registerAlertChannel(resource.id, `${name} incidentio`, this.program.generatedConstructFile(filename.fullPath));
    }
    gencode(logicalId, resource, context) {
        this.validateSafety(resource);
        const { id, file } = context.lookupAlertChannel(resource.id);
        file.namedImport(construct, 'checkly/constructs');
        const { config } = resource;
        file.section((0, sourcegen_1.decl)(id, builder => {
            builder.variable((0, sourcegen_1.expr)((0, sourcegen_1.ident)(construct), builder => {
                builder.new(builder => {
                    builder.string(logicalId);
                    builder.object(builder => {
                        builder.string('name', config.name);
                        builder.string('url', config.url);
                        if (config.headers) {
                            const apiKey = apiKeyFromHeaders(config.headers);
                            if (apiKey) {
                                builder.string('apiKey', apiKey);
                            }
                            else {
                                throw new Error(`Failed to extract incident.io API Key from webhook headers`);
                            }
                        }
                        if (config.template) {
                            if (config.template !== incidentio_alert_channel_1.IncidentioAlertChannel.DEFAULT_PAYLOAD) {
                                builder.string('payload', config.template);
                            }
                        }
                        (0, alert_channel_codegen_1.buildAlertChannelProps)(builder, resource);
                    });
                });
            }));
            builder.export();
        }));
    }
}
exports.IncidentioAlertChannelCodegen = IncidentioAlertChannelCodegen;
