// checkly/dist/constructs/multi-step-check-codegen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiStepCheckCodegen = void 0;
const sourcegen_1 = require("../sourcegen");
const playwright_config_codegen_1 = require("./playwright-config-codegen");
const check_codegen_1 = require("./check-codegen");
const codegen_1 = require("./internal/codegen");
const construct = 'MultiStepCheck';
class MultiStepCheckCodegen extends codegen_1.Codegen {
    describe(resource) {
        return `Multistep Check: ${resource.name}`;
    }
    gencode(logicalId, resource, context) {
        const filePath = context.filePath('resources/multi-step-checks', resource.name, {
            tags: resource.tags,
            isolate: true,
            unique: true,
        });
        const file = this.program.generatedConstructFile(filePath.fullPath);
        file.namedImport(construct, 'checkly/constructs');
        file.section((0, sourcegen_1.expr)((0, sourcegen_1.ident)(construct), builder => {
            builder.new(builder => {
                builder.string(logicalId);
                builder.object(builder => {
                    builder.object('code', builder => {
                        const scriptFile = this.program.staticSpecFile(filePath.extless, resource.script);
                        builder.string('entrypoint', file.relativePath(scriptFile));
                    });
                    if (resource.playwrightConfig) {
                        builder.value('playwrightConfig', (0, playwright_config_codegen_1.valueForPlaywrightConfig)(resource.playwrightConfig));
                    }
                    (0, check_codegen_1.buildCheckProps)(this.program, file, builder, resource, context);
                });
            });
        }));
    }
}
exports.MultiStepCheckCodegen = MultiStepCheckCodegen;
