// checkly/dist/constructs/private-location-codegen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivateLocationCodegen = exports.valueForPrivateLocationFromId = void 0;
const codegen_1 = require("./internal/codegen");
const sourcegen_1 = require("../sourcegen");
const construct = 'PrivateLocation';
function valueForPrivateLocationFromId(genfile, physicalId) {
    genfile.namedImport(construct, 'checkly/constructs');
    return (0, sourcegen_1.expr)((0, sourcegen_1.ident)(construct), builder => {
        builder.member((0, sourcegen_1.ident)('fromId'));
        builder.call(builder => {
            builder.string(physicalId);
        });
    });
}
exports.valueForPrivateLocationFromId = valueForPrivateLocationFromId;
class PrivateLocationCodegen extends codegen_1.Codegen {
    describe(resource) {
        return `Private Location: ${resource.name}`;
    }
    prepare(logicalId, resource, context) {
        const filePath = context.filePath('resources/private-locations', resource.slugName, {
            unique: true,
        });
        context.registerPrivateLocation(resource.id, resource.name, this.program.generatedConstructFile(filePath.fullPath));
    }
    gencode(logicalId, resource, context) {
        const { id, file } = context.lookupPrivateLocation(resource.id);
        file.namedImport(construct, 'checkly/constructs');
        file.section((0, sourcegen_1.decl)(id, builder => {
            builder.variable((0, sourcegen_1.expr)((0, sourcegen_1.ident)(construct), builder => {
                builder.new(builder => {
                    builder.string(logicalId);
                    builder.object(builder => {
                        builder.string('name', resource.name);
                        builder.string('slugName', resource.slugName);
                        if (resource.icon) {
                            builder.string('icon', resource.icon);
                        }
                        if (resource.proxyUrl) {
                            builder.string('proxyUrl', resource.proxyUrl);
                        }
                    });
                });
            }));
            builder.export();
        }));
    }
}
exports.PrivateLocationCodegen = PrivateLocationCodegen;
