// checkly/dist/constructs/status-page.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatusPage = void 0;
const construct_1 = require("./construct");
const project_1 = require("./project");
const ref_1 = require("./ref");
/**
 * Creates a Status Page
 */
class StatusPage extends construct_1.Construct {
    name;
    cards;
    url;
    customDomain;
    logo;
    redirectTo;
    favicon;
    defaultTheme;
    static __checklyType = 'status-page';
    /**
     * Constructs the Status Page instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props status page configuration properties
     *
     * {@link https://checklyhq.com/docs/cli/constructs-reference/#statuspage Read more in the docs}
     */
    constructor(logicalId, props) {
        super(StatusPage.__checklyType, logicalId);
        this.name = props.name;
        this.url = props.url;
        this.cards = props.cards;
        this.customDomain = props.customDomain;
        this.logo = props.logo;
        this.redirectTo = props.redirectTo;
        this.favicon = props.favicon;
        this.defaultTheme = props.defaultTheme;
        project_1.Session.registerConstruct(this);
    }
    synthesize() {
        return {
            name: this.name,
            url: this.url,
            customDomain: this.customDomain,
            logo: this.logo,
            redirectTo: this.redirectTo,
            favicon: this.favicon,
            defaultTheme: this.defaultTheme,
            cards: this.cards.map(card => ({
                name: card.name,
                services: card
                    .services
                    ?.map((service) => ref_1.Ref.from(service.logicalId)),
            })),
        };
    }
}
exports.StatusPage = StatusPage;
