// checkly/dist/constructs/telegram-alert-channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TelegramAlertChannel = void 0;
const webhook_alert_channel_1 = require("./webhook-alert-channel");
/**
 * Creates a Telegram Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
class TelegramAlertChannel extends webhook_alert_channel_1.WebhookAlertChannel {
    static DEFAULT_PAYLOAD = `<b>{{ALERT_TITLE}}</b> at {{STARTED_AT}} in {{RUN_LOCATION}} ({{RESPONSE_TIME}}ms)
Tags: {{#each TAGS}} <i><b>{{this}}</b></i> {{#unless @last}},{{/unless}} {{/each}}
<a href="{{RESULT_LINK}}">View check result</a>
`;
    /**
       * Constructs the Telegram Alert Channel instance
       *
       * @param logicalId unique project-scoped resource name identification
       * @param props Telegram alert channel configuration properties
       * Fix following url:
       * {@link https://checklyhq.com/docs/cli/constructs/#telegramalertchannel Read more in the docs}
      */
    constructor(logicalId, props) {
        // @ts-ignore
        super(logicalId, props);
        this.webhookType = 'WEBHOOK_TELEGRAM';
        this.method = 'POST';
        const payload = props.payload ?? TelegramAlertChannel.DEFAULT_PAYLOAD;
        // For historical reasons the payload is not escaped even though it
        // should be.
        this.template = `chat_id=${props.chatId}&parse_mode=HTML&text=${payload}`;
        this.url = `https://api.telegram.org/bot${props.apiKey}/sendMessage`;
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'WEBHOOK',
            config: {
                name: this.name,
                webhookType: this.webhookType,
                url: this.url,
                template: this.template,
                method: this.method,
                headers: this.headers,
                queryParameters: this.queryParameters,
                webhookSecret: this.webhookSecret,
            },
        };
    }
}
exports.TelegramAlertChannel = TelegramAlertChannel;
