// checkly/dist/services/check-parser/package-files/source-file.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceFile = exports.FileMeta = void 0;
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
class FileMeta {
    filePath;
    dirname;
    basename;
    constructor(filePath, dirname, basename) {
        this.filePath = filePath;
        this.dirname = dirname;
        this.basename = basename;
    }
    static fromFilePath(filePath) {
        return new FileMeta(filePath, node_path_1.default.dirname(filePath), node_path_1.default.basename(filePath));
    }
}
exports.FileMeta = FileMeta;
class SourceFile {
    static #id = 0;
    id = ++SourceFile.#id;
    contents;
    meta;
    constructor(meta, contents) {
        this.meta = meta;
        this.contents = contents;
    }
    static loadFromFilePath(filePath) {
        try {
            const contents = node_fs_1.default.readFileSync(filePath, {
                encoding: 'utf8',
            });
            const meta = FileMeta.fromFilePath(filePath);
            return new SourceFile(meta, contents);
        }
        catch (err) {
        }
    }
}
exports.SourceFile = SourceFile;
