// checkly/dist/loader/native.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NativeFileLoader = exports.hasNativeTypeScriptSupport = exports.detectNativeTypeScriptSupport = exports.DenoDetector = exports.BunDetector = void 0;
const package_manager_1 = require("../services/check-parser/package-files/package-manager");
Object.defineProperty(exports, "BunDetector", { enumerable: true, get: function () { return package_manager_1.BunDetector; } });
Object.defineProperty(exports, "DenoDetector", { enumerable: true, get: function () { return package_manager_1.DenoDetector; } });
const util_1 = require("../services/util");
const loader_1 = require("./loader");
const match_1 = require("./match");
function detectNativeTypeScriptSupport() {
    if (new package_manager_1.BunDetector().detectRuntime()) {
        return true;
    }
    if (new package_manager_1.DenoDetector().detectRuntime()) {
        return true;
    }
    return false;
}
exports.detectNativeTypeScriptSupport = detectNativeTypeScriptSupport;
exports.hasNativeTypeScriptSupport = detectNativeTypeScriptSupport();
class NativeFileLoader extends loader_1.FileLoader {
    constructor(options) {
        super({
            match: exports.hasNativeTypeScriptSupport
                ? match_1.FileMatch.any()
                : match_1.FileMatch.standardFiles(),
            ...options,
        });
    }
    async loadFile(filePath) {
        const moduleExports = await import((0, util_1.pathToPosix)(filePath));
        return moduleExports;
    }
}
exports.NativeFileLoader = NativeFileLoader;
