// checkly/dist/sourcegen/arraybuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayValueBuilder = exports.array = void 0;
const objectbuilder_1 = require("./objectbuilder");
const array_1 = require("./array");
const boolean_1 = require("./boolean");
const null_1 = require("./null");
const number_1 = require("./number");
const string_1 = require("./string");
const exprbuilder_1 = require("./exprbuilder");
const undefined_1 = require("./undefined");
const identifier_1 = require("./identifier");
function array(build) {
    const builder = new ArrayValueBuilder();
    build(builder);
    return builder.build();
}
exports.array = array;
class ArrayValueBuilder {
    #elements = [];
    empty() {
        this.#elements = [];
        return this;
    }
    string(value) {
        try {
            return this.value(new string_1.StringValue(value));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (string): ${cause}`, { cause });
        }
    }
    boolean(value) {
        try {
            return this.value(new boolean_1.BooleanValue(value));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (boolean): ${cause}`, { cause });
        }
    }
    number(value) {
        try {
            return this.value(new number_1.NumberValue(value));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (number): ${cause}`, { cause });
        }
    }
    null() {
        try {
            return this.value(new null_1.NullValue());
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (null): ${cause}`, { cause });
        }
    }
    undefined() {
        try {
            return this.value(new undefined_1.UndefinedValue());
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (undefined): ${cause}`, { cause });
        }
    }
    ident(value) {
        try {
            return this.value(new identifier_1.IdentifierValue(value));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (identifier): ${cause}`, { cause });
        }
    }
    array(build) {
        try {
            return this.value(array(build));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (array): ${cause}`, { cause });
        }
    }
    object(build) {
        try {
            return this.value((0, objectbuilder_1.object)(build));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (object): ${cause}`, { cause });
        }
    }
    expr(context, build) {
        try {
            return this.value((0, exprbuilder_1.expr)(context, build));
        }
        catch (cause) {
            throw new Error(`Failed to add array item #${this.#elements.length} (expr): ${cause}`, { cause });
        }
    }
    value(value) {
        this.#elements.push(value);
        return this;
    }
    get elements() {
        return this.#elements;
    }
    build() {
        return new array_1.ArrayValue([...this.#elements]);
    }
}
exports.ArrayValueBuilder = ArrayValueBuilder;
