// checkly/dist/sourcegen/declbuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeclarationBuilder = exports.decl = void 0;
const decl_1 = require("./decl");
function decl(identifier, build) {
    const builder = new DeclarationBuilder(identifier);
    build(builder);
    return builder.build();
}
exports.decl = decl;
class DeclarationBuilder {
    #identifier;
    #decl;
    constructor(identifier) {
        this.#identifier = identifier;
        this.#decl = new decl_1.IdentifierDeclaration(identifier);
    }
    variable(value, options) {
        this.#decl = new decl_1.VariableDeclaration(this.#identifier, value, options);
        return this;
    }
    export(options) {
        this.#decl = new decl_1.ExportDeclaration(this.#decl, options);
        return this;
    }
    leadingComment(comment) {
        this.#decl = new decl_1.LeadingCommentDeclaration(comment, this.#decl);
        return this;
    }
    trailingComment(comment) {
        this.#decl = new decl_1.TrailingCommentDeclaration(comment, this.#decl);
        return this;
    }
    build() {
        return this.#decl;
    }
}
exports.DeclarationBuilder = DeclarationBuilder;
